certbot/certbot/plugins/common.py

```diff
@@ -209,37 +219,44 @@ def rollback_checkpoints(self, rollback=1):
             raise errors.PluginError(str(err))
 
     @property
-    def ssl_dhparams(self):
+    def ssl_dhparams(self) -> str:
         """Full absolute path to ssl_dhparams file."""
         return os.path.join(self.config.config_dir, constants.SSL_DHPARAMS_DEST)
 
     @property
-    def updated_ssl_dhparams_digest(self):
+    def updated_ssl_dhparams_digest(self) -> str:
         """Full absolute path to digest of updated ssl_dhparams file."""
         return os.path.join(self.config.config_dir, constants.UPDATED_SSL_DHPARAMS_DIGEST)
 
-    def install_ssl_dhparams(self):
+    def install_ssl_dhparams(self) -> None:
         """Copy Certbot's ssl_dhparams file into the system's config dir if required."""
-        return install_version_controlled_file(
+        install_version_controlled_file(
             self.ssl_dhparams,
             self.updated_ssl_dhparams_digest,
             constants.SSL_DHPARAMS_SRC,
             constants.ALL_SSL_DHPARAMS_HASHES)
 
 
+class Configurator(Installer, interfaces.Authenticator, metaclass=ABCMeta):
```

Comment:
The concept of type intersection does not exist in mypy and other type checkers. Maybe because there is usually two alternatives that better fits the same purpose:
1) either you formally want to merge the functional logic of two parents class: then you can use the Python multi-inheritance system (and so use the dedicated class hierarchy formed by this inheritance as types)
2) or you want to structurally validate the type with duck typing: then you can define a `Protocol` that defines the contract of methods a class must conform with, and use this protocol as type (eg. `Iterable` with `__iter__()` or `Sized` with `__len__()`)

Given that two classes (`ApacheConfigurator` and `NginxConfigurator`) effectively merge the functionalities of `interfaces.Authenticator`  and `common.Installer`, I think it is worth to follow 1) and define a dedicated abstract class that does that, and use it as type in the `ChallengePerformer`.